### PR TITLE
 Add the possibility to change the color of individual events #103 

### PR DESCRIPTION
**Pull request by KebabRonin**
Added the possibility to edit the color of individual events within a calendar.

Each event now has a `Background Color` and a `Text Color`. When these fields are empty they default to the colors defined in the parent Calendar.

image

image

image

### CODE REVIEW

**Comment by KebabRonin** [comment]
**Known Issue:**
* The colorPicker remains locked in place when scrolling the page, breaking the UI

image

File: application-mocca-calendar-ui/src/main/resources/MoccaCalendar/Code/Macro.xml [hunk 1]
```diff
@@ -606,6 +606,10 @@ if (typeof(XWiki) == "undefined" || typeof(XWiki.widgets) == "undefined" || type
         script.remove();
       });
       content.insert('&lt;div id="formcontent"&gt;&lt;div id="forminnercontent"&gt;' + container.innerHTML + '&lt;/div&gt;&lt;/div&gt;');
+
+      //trigger xwiki:dom:updated right after loading xform in modal
+      var parentModalPopup = jQuery('.xdialog-box-moccacal-modal-popup').toArray();
+      document.fire('xwiki:dom:updated', {elements: parentModalPopup});
```
Comment (KebabRonin) [inline application-mocca-calendar-ui/src/main/resources/MoccaCalendar/Code/Macro.xml:612]: Event fired right after the popup becomes part of the document DOM.

Used `document.fire` instead of `$(document).trigger` to be consistent with the rest of the file.

File: application-mocca-calendar-api/src/main/java/org/xwiki/contrib/moccacalendar/script/MoccaCalendarScriptService.java [hunk 2]
```diff
@@ -618,12 +622,21 @@ private void fillInColorsFromNearestCalendar(EventInstance event)
...
             if (calendarData == null) {
                 // some arbitrary defaults
-                event.setBackgroundColor("#888");
+                if (event.getBackgroundColor().length() == 0) {
+                    event.setBackgroundColor("#888");
+                }
                 // text color can be missing
-                event.setTextColor("");
+                if (event.getTextColor().length() == 0) {
+                    // not that great
```
Comment (oanalavinia) [inline application-mocca-calendar-api/src/main/java/org/xwiki/contrib/moccacalendar/script/MoccaCalendarScriptService.java:630]: This comment is not needed

File: application-mocca-calendar-ui/src/main/resources/MoccaCalendar/Code/Macro.xml [hunk 3]
```diff
@@ -606,6 +606,10 @@ if (typeof(XWiki) == "undefined" || typeof(XWiki.widgets) == "undefined" || type
         script.remove();
       });
       content.insert('&lt;div id="formcontent"&gt;&lt;div id="forminnercontent"&gt;' + container.innerHTML + '&lt;/div&gt;&lt;/div&gt;');
+
+      //trigger xwiki:dom:updated right after loading xform in modal
```
Comment (oanalavinia) [inline application-mocca-calendar-ui/src/main/resources/MoccaCalendar/Code/Macro.xml:610]: This comment is not really needed, since I find the code below descriptive enough.
But if you feel it is important, it could be something like `Notify that the calendar modal was added in the document content`, in order to not simply duplicate the code below

File: application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml [hunk 4]
```diff
@@ -117,6 +117,7 @@ $doc.display($field)
   #end
 #end
 {{/velocity}}{{velocity}}{{html wiki="true" clean="false"}}
+#template('display_macros.vm')## Comment to escape new line so the UI displays properly
```
Comment (oanalavinia) [inline application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml:120]: The comment is not actually needed, since it means we should add this comment every time we use `##` like this

File: application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml [hunk 5]
```diff
@@ -236,6 +267,12 @@ $doc.display($field)
   &lt;/div&gt;
   #end ## if isSingleInstance
 
+    ##
+    ## Load required js/css for colorPicker when included/modal in another page
```
Comment (oanalavinia) [inline application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml:271]: All comments should end with a dot `.`

File: application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml [hunk 6]
```diff
@@ -266,6 +318,8 @@ $doc.display($field)
   ## This branch is only reached by events from iCals and gCals. The "xwiki false" setting is necessary because we
   ## don't have control over the data from other providers. If the wiki syntax is set to true, the data might be
   ## interpreted in a strange way and break the display.
+
+  ## (event text and bg colors might be broken for this branch)
```
Comment (oanalavinia) [inline application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml:322]: This comment shouldn't be added here. First of all, why would them be broken?

File: application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml [hunk 7]
```diff
@@ -256,6 +293,21 @@ $doc.display($field)
     : $doc.display('allDay')
     ; &lt;label for="description"&gt;$escapetool.xml($doc.displayPrettyName('description', false, false))&lt;/label&gt;
     : $eventInstance.getDescriptionHtml()
+      ## The colors here are for the XWiki Object (from the page url).
+      ## Colors displayed in the calendar are defined in the MoccaCalendarScriptService.java file.
+      ## There is another validation check made in MoccaCalendarScriptService.java. See the fillInColorsFromNearestCalendar() method.
+    ; &lt;label for="backgroundColor"&gt;$escapetool.xml($doc.displayPrettyName('backgroundColor', false, false))&lt;/label&gt;
+    #if ($doc.getValue('backgroundColor').length() == 0) ## If the color is valid or should default to parent calendar colors.
```
Comment (oanalavinia) [inline application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml:300]: Aren't you already defaulting on the calendar value in case this is not defined? Why you need to do this again in here?

File: application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml [hunk 8]
```diff
@@ -256,6 +293,21 @@ $doc.display($field)
     : $doc.display('allDay')
     ; &lt;label for="description"&gt;$escapetool.xml($doc.displayPrettyName('description', false, false))&lt;/label&gt;
     : $eventInstance.getDescriptionHtml()
+      ## The colors here are for the XWiki Object (from the page url).
+      ## Colors displayed in the calendar are defined in the MoccaCalendarScriptService.java file.
+      ## There is another validation check made in MoccaCalendarScriptService.java. See the fillInColorsFromNearestCalendar() method.
```
Comment (oanalavinia) [inline application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml:298]: This comment is not very helpful. Some things related to understanding the code should be left to be discovered by just looking at the code when it's needed, in order to not make the code harder to read. For this example specific I feel like it would be also hard to follow, since in case the code changes, this comment will remain outdated, doing more harm than good
You should add a comment if you feel some things are harder to be spotted, or something complex is involved

**Review by oanalavinia — CHANGES_REQUESTED** [review]

File: application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml (hunk 7 above)
Comment (KebabRonin) [reply application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml:300]: This is only for the View page, which is not directly part of the calendar (it can be viewed as a standalone page).

In the case when the color is not set, the color displayed in the UI would be black. This #if makes it so the color from the parent calendar is displayed instead, with the "(Calendar Default)" text. For invalid colors (like "z") the color displayed in the View page is still black.

The css in the calendar macro normally overrides the css of the event (if the color is invalid) so the color is a prettier default (blue and white instead of black and black). 

View page without if:
image
View page with if:
image

File: application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml [hunk 9]
```diff
@@ -228,14 +244,33 @@ $doc.display($field)
     : #displayDate('endDate')#showvalidationmessage("MoccaCalendar.Code.MoccaCalendarEventModificationClass.validation.endDateBeforeStartDate")#showvalidationmessage("MoccaCalendar.Code.MoccaCalendarEventModificationClass.validation.endDateNotDate")
     ; &lt;label for="${editOnePrefix}_description"&gt;$escapetool.xml($doc.displayPrettyName('description', false, false))&lt;/label&gt;
     : $doc.display('description')
-
+    ; &lt;label for="${editOnePrefix}_backgroundColor"&gt;$escapetool.xml($doc.displayPrettyName('backgroundColor', false, false))&lt;/label&gt;
+    #set ($colorPickerParams = {
+      'id': "${editOnePrefix}_backgroundColor",
+      'name': "${editOnePrefix}_backgroundColor",
+      'value': $escapetool.xml($!doc.getValue('backgroundColor'))
```
Comment (oanalavinia) [inline application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml:251]: Same

File: application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml [hunk 10]
```diff
@@ -228,14 +244,33 @@ $doc.display($field)
...
+    })
+    : #colorPicker($colorPickerParams)
+    ; &lt;label for="${editOnePrefix}_textColor"&gt;$escapetool.xml($doc.displayPrettyName('textColor', false, false))&lt;/label&gt;
+    #set ($colorPickerParams = {
+      'id': "${editOnePrefix}_textColor",
+      'name': "${editOnePrefix}_textColor",
+      'value': $escapetool.xml($!doc.getValue('textColor'))
```
Comment (oanalavinia) [inline application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml:258]: Same

File: application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml [hunk 11]
```diff
@@ -184,6 +186,20 @@ $doc.display($field)
     : $doc.display('allDay')
     ; &lt;label for="${editPrefix}_description"&gt;$escapetool.xml($doc.displayPrettyName('description', false, false))&lt;/label&gt;
     : $doc.display('description')
+    ; &lt;label for="${editPrefix}_backgroundColor"&gt;$escapetool.xml($doc.displayPrettyName('backgroundColor', false, false))&lt;/label&gt;
+    #set ($colorPickerParams = {
+      'id': "${editPrefix}_backgroundColor",
+      'name': "${editPrefix}_backgroundColor",
+      'value': $escapetool.xml($!doc.getValue('backgroundColor'))
```
Comment (oanalavinia) [inline application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml:193]: Why don't you use `$eventInstance.getBackgroundColor()`? You can initialize the `eventInstance` object so it can be accessed here as well

File: application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml [hunk 12]
```diff
@@ -184,6 +186,20 @@ $doc.display($field)
...
+    })
+    : #colorPicker($colorPickerParams)
+    ; &lt;label for="${editPrefix}_textColor"&gt;$escapetool.xml($doc.displayPrettyName('textColor', false, false))&lt;/label&gt;
+    #set ($colorPickerParams = {
+      'id': "${editPrefix}_textColor",
+      'name': "${editPrefix}_textColor",
+      'value': $escapetool.xml($!doc.getValue('textColor'))
```
Comment (oanalavinia) [inline application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml:200]: Why don't you use `$eventInstance.getTextColor()`?

File: application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml (hunk 7 above)
Comment (oanalavinia) [reply application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml:295]: You should simply use `$eventInstance.getTextColor()/getBackgroundColor()`, which will give you the right color (from event or parent calendar). I find it ok even if it means not displaying the `Calendar default` info (but could also be done)
Practically you would have something similar to: 
```
#set ($textColor = $eventInstance.getTextColor())
#set (isCalendarDefault = ($textColor == $calendarDoc.getValue('color')) )
```
And then use the `$isCalendarDefault` variable in order to display or not the `(Calendar default)` message
Not that this message should be added using a [translation key](https://extensions.xwiki.org/xwiki/bin/view/Extension/Localization/)

**Review by oanalavinia — CHANGES_REQUESTED** [review]

File: application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml (hunk 11 above)
Comment (KebabRonin) [reply application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml]: When using ```$eventInstance.getBackgroundColor()```, the final displayed color is returned, perhaps replaced by the calendar color.

This means that when editing, the value of the field can be set to the current calendar default, overriding the actual value of the event color when saved. As a result, the color of an event will remain the same, not changing when the calendar default changes, even if it wasn't explicitly assigned a color by the user.

File: application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml [hunk 13]
```diff
@@ -131,7 +133,12 @@ $doc.display($field)
       #end
     {{/error}}
   #end
+  #if(!$doc.isNew())
+    #set($eventInstance = $services.moccacalendar.getEventInstance($doc, $instanceStart))
+    #set($eventInstanceRecurrentBase = $services.moccacalendar.getEventInstance($doc, null))
```
Comment (ChiuchiuSorin) [inline application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml:138]: There should be a space between `set` and `(`

File: application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml [hunk 14]
```diff
@@ -291,6 +340,14 @@ $doc.display($field)
       &lt;dd&gt;
         $request.description
       &lt;/dd&gt;
+      &lt;dt&gt;&lt;label for="backgroundColor"&gt;$escapetool.xml($services.localization.render('MoccaCalendar.MoccaCalendarEventClass_backgroundColor'))&lt;/label&gt;&lt;/dt&gt;
```
Comment (oanalavinia) [inline application-mocca-calendar-ui/src/main/resources/MoccaCalendar/MoccaCalendarEventSheet.xml:343]: This line is too long (>120 characters)
The idea of keeping lines in the 120ch limit is to have it in your computer visual limit, so you don't have to scroll horizontally when looking at the code
Below as well

**Review by oanalavinia — CHANGES_REQUESTED** [review]

**Review by oanalavinia — APPROVED** [review]
